libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -241,36 +241,74 @@ void GCS_MAVLINK::send_battery_status(const uint8_t instance) const
     bool got_temperature = battery.get_temperature(temp, instance);
 
     // prepare arrays of individual cell voltages
-    uint16_t cell_volts[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN];
-    uint16_t cell_volts_ext[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_EXT_LEN];
+    uint16_t cell_mvolts[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN];
+    uint16_t cell_mvolts_ext[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_EXT_LEN];
+    const uint16_t max_cell_mV = 65534;
+
     if (battery.has_cell_voltages(instance)) {
         const AP_BattMonitor::cells& batt_cells = battery.get_cell_voltages(instance);
+        static_assert(sizeof(cell_mvolts) <= sizeof(batt_cells.cells), "cell array length not large enough");
+
         // copy the first 10 cells
-        memcpy(cell_volts, batt_cells.cells, sizeof(cell_volts));
+        memcpy(cell_mvolts, batt_cells.cells, sizeof(cell_mvolts));
         // 11 ... 14 use a second cell_volts_ext array
         for (uint8_t i = 0; i < MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_EXT_LEN; i++) {
             if (MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN+i < uint8_t(ARRAY_SIZE(batt_cells.cells))) {
-                cell_volts_ext[i] = batt_cells.cells[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN+i];
+                cell_mvolts_ext[i] = batt_cells.cells[MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN+i];
             } else {
-                cell_volts_ext[i] = 0;
+                cell_mvolts_ext[i] = 0;
+            }
+        }
+        /*
+          now adjust voltages to cope with two things:
+             1) we may be reporting sag corrected voltage
+             2) the battery may have more cells than can be reported by the backend, so the actual voltage may be higher than the sum
+        */
+        const float voltage_mV = battery.gcs_voltage(instance) * 1e3f;
+        float voltage_mV_sum = 0;
+        uint8_t non_zero_cell_count = 0;
+        for (uint8_t i=0; i<MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN; i++) {
+            if (cell_mvolts[i] > 0) {
```

Comment:
These will almost never be 0. An unknown cell is stored as `0xFFFF`.

Reply:
right, almost never zero, but it can be zero, and I wanted to handle that

Reply:
Sorry, I should be more clear on the concern I have. A cell that isn't monitored will be `0xFFFF` so on a 4S pack like Solo for example the last 8 cells will all be non 0, and you will accumulate a ton of extra voltage that isn't actually in the system. I believe this causes an error when you try to deal with extra voltage that may need to be distributed below.

Reply:
ok, I see, I had thought the 65535 magic was only for mavlink, I now see SMBUS also does it, so I've changed this code to not include any 65535 values in the sum